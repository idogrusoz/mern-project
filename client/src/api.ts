import axios from "axios";

export default axios.create({
    baseURL: process.env.NODE_ENV === "production" ? process.env.PUBLIC_URL : `http://localhost:5000/api/v1/`,
    withCredentials: true,
});
