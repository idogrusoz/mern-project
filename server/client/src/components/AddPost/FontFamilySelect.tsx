import { MenuItem, Select } from "@material-ui/core";
import React, { Dispatch, FunctionComponent, SetStateAction } from "react";

type FontFamilySelectProps = {
    setFontFamily: Dispatch<SetStateAction<string>>;
    setFontWeightValues: Dispatch<SetStateAction<Array<number>>>;
};

interface fontInfo {
    label: string;
    value: string;
    options: number[];
}

const FontFamilySelect: FunctionComponent<FontFamilySelectProps> = ({ setFontFamily, setFontWeightValues }) => {
    const fontFamilies: fontInfo[] = [
        {
            label: "Caveat",
            value: "Caveat",
            options: [400, 700],
        },
        {
            label: "Dancing Script",
            value: "Dancing Script",
            options: [400, 700],
        },
        {
            label: "Kalam",
            value: "Kalam",
            options: [300, 400, 700],
        },
        {
            label: "Lemonada",
            value: "Lemonada",
            options: [300, 400, 700],
        },
        {
            label: "Lobster",
            value: "Lobster",
            options: [400, 700],
        },
        {
            label: "Montserrat",
            value: "Montserrat",
            options: [100, 400, 700, 900],
        },
        {
            label: "Montserrat Alternates",
            value: "Montserrat Alternates",
            options: [100, 400, 700, 900],
        },
        {
            label: "Permanent Marker",
            value: "Permanent Marker",
            options: [400],
        },
        {
            label: "Roboto",
            value: "Roboto",
            options: [100, 400, 900],
        },
        {
            label: "Sansita Swashed",
            value: "Sansita Swashed",
            options: [400, 700, 900],
        },
    ];

    const handleChange = (index: number) => {
        setFontFamily(fontFamilies[index].value);
        setFontWeightValues(fontFamilies[index].options);
    };

    return (
        <Select fullWidth defaultValue={8} onChange={(e) => handleChange(Number(e.target.value))}>
            {fontFamilies.map((font: fontInfo, i: number) => {
                return (
                    <MenuItem value={i} key={i}>
                        {font.label}
                    </MenuItem>
                );
            })}
        </Select>
    );
};

export default FontFamilySelect;
