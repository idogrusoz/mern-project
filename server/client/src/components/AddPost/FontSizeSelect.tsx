import { MenuItem, Select } from "@material-ui/core";
import React, { Dispatch, FunctionComponent, SetStateAction } from "react";

type FontSizeSelectProps = {
    fontSize: number;
    setFontSize: Dispatch<SetStateAction<number>>;
};

const FontSizeSelect: FunctionComponent<FontSizeSelectProps> = ({ fontSize, setFontSize }) => {
    const fontSizes: number[] = [16, 20, 22, 24, 28, 32, 36, 40, 48, 56, 64, 72, 144];
    return (
        <Select fullWidth defaultValue={fontSize} onChange={(e) => setFontSize(Number(e.target.value))}>
            {fontSizes.map((size: number, i: number) => {
                return (
                    <MenuItem value={size} key={i}>
                        {size}
                    </MenuItem>
                );
            })}
        </Select>
    );
};

export default FontSizeSelect;
