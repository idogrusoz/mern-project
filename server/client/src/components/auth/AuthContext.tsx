import React, {
    createContext,
    FunctionComponent,
    Dispatch,
    useMemo,
    useState,
    SetStateAction,
    ReactNode,
    useCallback,
} from "react";
import api from "../../api";
import { ResponseUser } from "../../../../interfaces/user.interfaces";

type AuthContextObject = {
    // isAuthenticated: () => Promise<void>;
    authenticated: boolean;
    setAuthenticated: Dispatch<SetStateAction<boolean>>;
    user: ResponseUser | null;
    setUser: Dispatch<SetStateAction<ResponseUser | null>>;
    logOut: () => Promise<void>;
};

type Props = {
    children: ReactNode;
};

const defaultCotextValue: AuthContextObject = {
    // isAuthenticated: () => new Promise(() => false),
    authenticated: false,
    setAuthenticated: () => {},
    user: null,
    setUser: () => {},
    logOut: () => new Promise(() => {}),
};

const AuthContext = createContext<AuthContextObject>(defaultCotextValue);

const AuthContextProvider: FunctionComponent<Props> = ({ children }) => {
    const [authenticated, setAuthenticated] = useState<boolean>(false);
    const [user, setUser] = useState<ResponseUser | null>(null);

    // Tobe uncommented once API route s built
    // const isAuthenticated = useCallback(async (): Promise<void> => {
    //     try {
    //         const response = await api.get("auth/authenticated");
    //         console.log(response.status);
    //         setAuthenticated(response.data.isAuthenticated);
    //         if (!user) {
    //             setUser(response.data.user);
    //         }
    //     } catch (error) {
    //         console.log(error);
    //         setAuthenticated(false);
    //         setUser(null);
    //     }
    // }, [user]);

    // useEffect(() => {
    //     const fetch = () => {
    //         isAuthenticated();
    //     };
    //     fetch();
    // }, [isAuthenticated]);

    const logOut = useCallback(async () => {
        try {
            await api.get("auth/logout");
        } catch (err) {
            console.log("Error : ", err);
            throw err;
        }
    }, []);

    const AuthContextValue: AuthContextObject = useMemo(() => {
        const value = {
            // isAuthenticated,
            authenticated,
            setAuthenticated,
            user,
            setUser,
            logOut,
        };
        return value;
    }, [
        // isAuthenticated,
        authenticated,
        setAuthenticated,
        user,
        setUser,
        logOut,
    ]);

    return <AuthContext.Provider value={AuthContextValue}>{children}</AuthContext.Provider>;
};

export { AuthContextProvider, AuthContext };
