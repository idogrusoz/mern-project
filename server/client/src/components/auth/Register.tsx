import { Button, Grid, makeStyles, TextField } from "@material-ui/core";
import { useFormik } from "formik";
import * as Yup from "yup";
import React, { Dispatch, FunctionComponent, SetStateAction, useState } from "react";
import api from "../../api";
import { UserInterface } from "../../../../interfaces/user.interfaces";

export type RegisterProps = {
    register: (user: UserInterface) => void;
    setShowSignin: Dispatch<SetStateAction<boolean>>;
};

const useStyles = makeStyles({
    container: {
        overflowX: "auto",
        minHeight: " 100vh",
    },
});

const Register: FunctionComponent<RegisterProps> = ({ register, setShowSignin }) => {
    const [userNameInput, setUserNameInput] = useState<string>("");
    const classes = useStyles();
    const formik = useFormik({
        initialValues: { firstName: "", lastName: "", userName: "", email: "", password: "", password2: "" },
        onSubmit: () => {
            const user: UserInterface = {
                ...formik.values,
                displayName: formik.values.firstName + " " + formik.values.lastName,
            };
            register(user);
        },
        validationSchema: Yup.object({
            firstName: Yup.string()
                .min(2, "Must be at least two chaeacters")
                .max(30, "Must be maximum 30 characters or less")
                .required("Required")
                .trim(),
            lastName: Yup.string()
                .min(2, "Must be at least two chaeacters")
                .max(30, "Must be maximum 30 characters or less")
                .required("Required")
                .trim(),
            userName: Yup.string()
                .min(3, "Must be at least three chaeacters")
                .max(30, "Must be maximum 30 characters or less")
                .required("Required")
                .test("check user name", "This username is not available", async (value: string | null | undefined) => {
                    if (value && value !== userNameInput) {
                        setUserNameInput(value);
                        return api
                            .post("users/username", { value })
                            .then((data) => data.data)
                            .catch((error) => false);
                    }
                    return value;
                })
                .trim(),
            password: Yup.string()
                .trim()
                // .matches(/^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9])(?=.*?[#?!@$%^&*-]).{8,}$/)
                .required(),
            password2: Yup.string()
                .trim()
                .oneOf([Yup.ref("password"), ""], "Passwords must match")
                .required(),
            email: Yup.string().email("Invalid email address").required("Required").trim(),
        }),
    });

    return (
        <form
            onSubmit={(e) => {
                e.preventDefault();
                formik.handleSubmit();
            }}
        >
            <Grid
                container
                justify="center"
                spacing={2}
                alignItems="center"
                direction="column"
                className={classes.container}
            >
                <Grid item>
                    <TextField
                        id="standard-basic"
                        label="First Name"
                        type="text"
                        name="firstName"
                        onChange={formik.handleChange}
                        value={formik.values.firstName}
                        error={formik.touched.firstName && formik.errors.firstName ? true : false}
                        helperText={formik.touched.firstName ? formik.errors.firstName : undefined}
                    />
                </Grid>
                <Grid item>
                    <TextField
                        id="standard-basic"
                        label="Last Name"
                        type="text"
                        name="lastName"
                        onChange={formik.handleChange}
                        value={formik.values.lastName}
                        error={formik.touched.lastName && formik.errors.lastName ? true : false}
                        helperText={formik.touched.lastName ? formik.errors.lastName : undefined}
                    />
                </Grid>
                <Grid item>
                    <TextField
                        id="standard-basic"
                        label="User Name"
                        type="text"
                        name="userName"
                        onChange={formik.handleChange}
                        value={formik.values.userName}
                        error={formik.touched.userName && formik.errors.userName ? true : false}
                        helperText={formik.touched.userName ? formik.errors.userName : undefined}
                    />
                </Grid>
                <Grid item>
                    <TextField
                        id="standard-basic"
                        label="Email"
                        type="text"
                        name="email"
                        onChange={formik.handleChange}
                        value={formik.values.email}
                        error={formik.touched.email && formik.errors.email ? true : false}
                        helperText={formik.touched.email ? formik.errors.email : undefined}
                    />
                </Grid>
                <Grid item>
                    <TextField
                        id="standard-basic"
                        label="Password"
                        type="Password"
                        name="password"
                        onChange={formik.handleChange}
                        value={formik.values.password}
                        error={formik.touched.password && formik.errors.password ? true : false}
                        helperText={
                            formik.touched.password
                                ? formik.errors.password
                                    ? "Invalid password"
                                    : undefined
                                : undefined
                        }
                    />
                </Grid>
                <Grid item>
                    <TextField
                        id="standard-basic"
                        label="Confirm Password"
                        type="Password"
                        name="password2"
                        onChange={formik.handleChange}
                        value={formik.values.password2}
                        error={formik.touched.password2 && formik.errors.password2 ? true : false}
                        helperText={formik.touched.password2 ? formik.errors.password2 : undefined}
                    />
                </Grid>
                <Grid item>
                    <Button type="submit" color="primary" variant="contained">
                        Register
                    </Button>
                </Grid>
                <Grid item>
                    <Button id="signinButton" color="primary" variant="contained" onClick={() => setShowSignin(true)}>
                        SignIn
                    </Button>
                </Grid>
            </Grid>
        </form>
    );
};

export default Register;
