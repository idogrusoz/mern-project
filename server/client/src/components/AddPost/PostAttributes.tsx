import { Button, Grid, makeStyles, useMediaQuery } from "@material-ui/core";
import React, { Dispatch, FunctionComponent, SetStateAction, useState } from "react";
import { useHistory } from "react-router-dom";
import ColorSelect from "./ColorSelect";
import FontFamilySelect from "./FontFamilySelect";
import FontSizeSelect from "./FontSizeSelect";
import FontWeightSelect from "./FontWeightSelect";
import TextAlignSelect from "./TextAlignSelect";
import TextContentInput from "./TextContentInput";

export type PostAttributesProps = {
    textContent: string;
    setTextContent: Dispatch<SetStateAction<string>>;
    fontFamily: string;
    setFontFamily: Dispatch<SetStateAction<string>>;
    color: string;
    setColor: Dispatch<SetStateAction<string>>;
    backgroundColor: string;
    setBackgroundColor: Dispatch<SetStateAction<string>>;
    fontWeight: number;
    setFontWeight: Dispatch<SetStateAction<number>>;
    fontSize: number;
    setFontSize: Dispatch<SetStateAction<number>>;
    textAlign: "left" | "right" | "justify" | "center";
    setTextAlign: Dispatch<SetStateAction<"left" | "right" | "justify" | "center">>;
    addPost: () => Promise<void>;
};

const PostAttributes: FunctionComponent<PostAttributesProps> = ({
    textContent,
    setTextContent,
    fontFamily,
    setFontFamily,
    color,
    setColor,
    backgroundColor,
    setBackgroundColor,
    fontWeight,
    setFontWeight,
    fontSize,
    setFontSize,
    textAlign,
    setTextAlign,
    addPost,
}) => {
    const [fontWeightValues, setFontWeightValues] = useState<Array<number>>([100, 400, 900]);
    const history = useHistory();
    const isBigScreen = useMediaQuery("(min-width: 650px)");
    const useStyles = makeStyles({
        container: isBigScreen
            ? {
                  maxWidth: "500px",
                  margin: "auto",
              }
            : { marginTop: 30 },
    });
    const classes = useStyles();
    return (
        <Grid container direction="row" spacing={5} className={classes.container}>
            <Grid container justify="space-around" direction="row">
                <Grid item xs={6}>
                    <Grid container justify="center">
                        <Grid item>
                            <Button color="primary" variant="contained" onClick={addPost}>
                                Share
                            </Button>
                        </Grid>
                    </Grid>
                </Grid>
                <Grid item xs={6}>
                    <Grid container justify="center">
                        <Grid item>
                            <Button color="secondary" variant="contained" onClick={() => history.goBack()}>
                                Cancel
                            </Button>
                        </Grid>
                    </Grid>
                </Grid>
            </Grid>
            <Grid item xs={12}>
                <TextContentInput textContent={textContent} setTextContent={setTextContent} />
            </Grid>
            <Grid item xs={6}>
                <FontFamilySelect setFontFamily={setFontFamily} setFontWeightValues={setFontWeightValues} />
            </Grid>
            <Grid item xs={6}>
                <FontWeightSelect
                    fontWeight={fontWeight}
                    setFontWeight={setFontWeight}
                    fontWeightValues={fontWeightValues}
                />
            </Grid>
            <Grid item xs={6}>
                <FontSizeSelect fontSize={fontSize} setFontSize={setFontSize} />
            </Grid>
            <Grid item xs={6}>
                <TextAlignSelect textAlign={textAlign} setTextAlign={setTextAlign} />
            </Grid>

            <Grid item lg={6} xs={12}>
                <ColorSelect label="Color" color={color} setColor={setColor} />
            </Grid>
            <Grid item lg={6} xs={12}>
                <ColorSelect label="Backgound Color" color={backgroundColor} setColor={setBackgroundColor} />
            </Grid>
        </Grid>
    );
};

export default PostAttributes;
