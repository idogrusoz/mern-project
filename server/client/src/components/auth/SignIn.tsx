import { Button, Grid, TextField } from "@material-ui/core";
import React, { useContext } from "react";
import { AuthContext } from "./AuthContext";

const SignIn = () => {
    const { email, setEmail, password, setPassword, setShowSignin, signin } = useContext(AuthContext);
    return (
        <>
            <Grid
                container
                justify="center"
                spacing={2}
                alignItems="center"
                direction="column"
                style={{ height: "100vh" }}
            >
                <Grid item>
                    <TextField
                        id="standard-basic"
                        label="Email"
                        type="Email"
                        onChange={(e: { target: { value: any } }) => setEmail(e.target.value)}
                        value={email}
                    />
                </Grid>
                <Grid item>
                    <TextField
                        id="standard-basic"
                        label="Password"
                        type="Password"
                        onChange={(e: { target: { value: any } }) => setPassword(e.target.value)}
                        value={password}
                    />
                </Grid>
                <Grid item>
                    <Button type="submit" variant="contained" color="primary" size="large" onClick={() => signin()}>
                        Sign In
                    </Button>
                </Grid>
                <Grid item>
                    <Button
                        type="submit"
                        variant="contained"
                        color="primary"
                        size="large"
                        onClick={() => {
                            console.log(1);
                            setShowSignin(false);
                        }}
                    >
                        Register
                    </Button>
                </Grid>
            </Grid>
        </>
    );
};

export default SignIn;
