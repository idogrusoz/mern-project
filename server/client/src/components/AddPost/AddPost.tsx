import { Grid, makeStyles, useMediaQuery } from "@material-ui/core";
import React, { useState, useContext } from "react";
import PostCanvas from "../Post/PostCanvas";
import PostAttributes from "./PostAttributes";
import { IPostInterface } from "../../../../interfaces/post.interfaces";
import { AuthContext } from "../Auth/AuthContext";
import api from "../../api";
import { useHistory } from "react-router-dom";

const AddPost = () => {
    const { user } = useContext(AuthContext);
    const history = useHistory();
    const [textContent, setTextContent] = useState<string>("");
    const [fontFamily, setFontFamily] = useState<string>("Roboto");
    const [color, setColor] = useState<string>("#000000");
    const [backgroundColor, setBackgroundColor] = useState<string>("#ffffff");
    const [fontWeight, setFontWeight] = useState<number>(400);
    const [fontSize, setFontSize] = useState<number>(16);
    const [textAlign, setTextAlign] = useState<"left" | "right" | "justify" | "center">("left");
    const [fontWeightValues, setFontWeightValues] = useState<Array<number>>([100, 400, 900]);
    const isBigScreen = useMediaQuery("(min-width: 650px)");
    const useStyles = makeStyles({
        container: {
            padding: isBigScreen ? "100px 120px" : "20px",
        },
    });
    const classes = useStyles();
    const addPost = async (): Promise<void> => {
        const post: IPostInterface = {
            author: {
                user_id: user!._id,
                userName: user!.userName,
                image: user!.image || "",
            },
            likes: [],
            textContent,
            style: {
                backgroundColor,
                color,
                fontFamily,
                fontSize,
                fontWeight,
                textAlign,
            },
        };
        try {
            await api.post("posts", post);
            history.goBack();
        } catch (error) {
            throw error;
        }
    };
    return (
        <Grid container justify="center" spacing={2} alignItems="center" direction="row" className={classes.container}>
            <Grid item md={12} lg={7}>
                <PostCanvas
                    backgroundColor={backgroundColor}
                    color={color}
                    fontFamily={fontFamily}
                    fontSize={fontSize}
                    fontWeight={fontWeight}
                    textAlign={textAlign}
                    textContent={textContent}
                />
            </Grid>
            <Grid item md={12} lg={5}>
                <PostAttributes
                    backgroundColor={backgroundColor}
                    color={color}
                    fontSize={fontSize}
                    fontWeight={fontWeight}
                    setBackgroundColor={setBackgroundColor}
                    setColor={setColor}
                    setFontFamily={setFontFamily}
                    setFontSize={setFontSize}
                    setFontWeight={setFontWeight}
                    setTextAlign={setTextAlign}
                    setTextContent={setTextContent}
                    textAlign={textAlign}
                    textContent={textContent}
                    addPost={addPost}
                    fontWeightValues={fontWeightValues}
                    setFontWeightValues={setFontWeightValues}
                />
            </Grid>
        </Grid>
    );
};

export default AddPost;
