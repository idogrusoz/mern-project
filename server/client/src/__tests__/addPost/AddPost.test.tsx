jest.mock("../../api.ts");
import React from "react";
import toJson from "enzyme-to-json";
import { mount, shallow } from "enzyme";
import api from "../../api";
import AddPost from "../../components/addPost/AddPost";
import { Button } from "@material-ui/core";
import { AuthContext, AuthContextObject, defaultContextValue } from "../../components/auth/AuthContext";
import { MemoryRouter } from "react-router-dom";

describe("AddPost test", () => {
    const value: AuthContextObject = {
        ...defaultContextValue,
        user: {
            _id: "userId",
            firstName: "firstName",
            lastName: "lastName",
            displayName: "displayName",
            userName: "userName",
            email: "email",
        },
        // isAuthenticated: jest.fn(() => Promise.resolve()),
        // authenticated: true,
        // setAuthenticated: jest.fn(() => {}),
        // setUser: jest.fn(() => {}),
        // signOut: jest.fn(() => Promise.resolve()),
    };
    const wrapper = mount(
        <AuthContext.Provider value={value}>
            <MemoryRouter>
                <AddPost />
            </MemoryRouter>
        </AuthContext.Provider>,
    );
    const mockApi = jest.spyOn(api, "post");
    const button = wrapper.find(Button).at(0);
    afterAll(() => {
        wrapper.unmount();
    });
    it("renders without crash", () => {
        const tree = shallow(<AddPost />);
        expect(toJson(tree)).toMatchSnapshot();
    });
    it("calls api when post is sent", () => {
        expect(button.text()).toEqual("Share");
        button.simulate("click");
        expect(mockApi).toBeCalled();
    });
});
