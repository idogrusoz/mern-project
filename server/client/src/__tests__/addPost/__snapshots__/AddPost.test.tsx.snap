// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddPost test renders without crash 1`] = `
<WithStyles(ForwardRef(Grid))
  alignItems="center"
  className="makeStyles-container-5"
  container={true}
  direction="row"
  justify="center"
  spacing={2}
>
  <WithStyles(ForwardRef(Grid))
    item={true}
    lg={7}
    md={12}
  >
    <PostCanvas
      backgroundColor="#ffffff"
      color="#000000"
      fontFamily="Roboto"
      fontSize={16}
      fontWeight={400}
      textAlign="left"
      textContent=""
    />
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    item={true}
    lg={5}
    md={12}
  >
    <PostAttributes
      addPost={[Function]}
      backgroundColor="#ffffff"
      color="#000000"
      fontSize={16}
      fontWeight={400}
      fontWeightValues={
        Array [
          100,
          400,
          900,
        ]
      }
      setBackgroundColor={[Function]}
      setColor={[Function]}
      setFontFamily={[Function]}
      setFontSize={[Function]}
      setFontWeight={[Function]}
      setFontWeightValues={[Function]}
      setTextAlign={[Function]}
      setTextContent={[Function]}
      textAlign="left"
      textContent=""
    />
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
